File: src/test/java/config/ConfigReaderCmTest.java
package config;

import annotations.Component;
import annotations.Layer;
import org.aeonbits.owner.ConfigFactory;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.parallel.Execution;
import org.junit.jupiter.api.parallel.ExecutionMode;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

@Component("cm")
@Layer("unit")
@DisplayName("ConfigReader CM URLs")
@Execution(ExecutionMode.SAME_THREAD)
class ConfigReaderCmTest {

    private static TestConfig configWith(Map<String, String> overrides) {
        return ConfigFactory.create(TestConfig.class, overrides);
    }

    @Test
    @DisplayName("resolveCmHubUrl uses cmHubPort with trailing slash")
    void resolveCmHubUrlUsesCmHubPort() {
        var config = configWith(Map.of("cmHubPort", "4445"));
        assertEquals("http://127.0.0.1:4445/", ConfigReader.resolveCmHubUrl(config));
    }

    @Test
    @DisplayName("resolveCmHubUrl respects custom cmHubPort")
    void resolveCmHubUrlRespectsCustomPort() {
        var config = configWith(Map.of("cmHubPort", "9444"));
        assertEquals("http://127.0.0.1:9444/", ConfigReader.resolveCmHubUrl(config));
    }

    @Test
    @DisplayName("resolveCmUiUrl uses cmUiPort with trailing slash")
    void resolveCmUiUrlUsesCmUiPort() {
        var config = configWith(Map.of("cmUiPort", "8081"));
        assertEquals("http://127.0.0.1:8081/", ConfigReader.resolveCmUiUrl(config));
    }

    @Test
    @DisplayName("resolveCmUiUrl respects custom cmUiPort")
    void resolveCmUiUrlRespectsCustomPort() {
        var config = configWith(Map.of("cmUiPort", "9080"));
        assertEquals("http://127.0.0.1:9080/", ConfigReader.resolveCmUiUrl(config));
    }

    @Test
    @DisplayName("resolveCmRemoteUrl points WebDriver hub at CM-managed port")
    void resolveCmRemoteUrlCombinesHubBaseAndWdHubPath() {
        var config = configWith(Map.of("cmHubPort", "4445"));
        assertEquals("http://127.0.0.1:4445/wd/hub", ConfigReader.resolveCmRemoteUrl(config));
    }
}

File: internal/config/config_reader_cm_test.go
package config_test

import (
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/qa-guru/selenoid-tests/internal/allurex"
	"github.com/qa-guru/selenoid-tests/internal/config"
)

func TestConfigReader_ResolveCmHubURLUsesCmHubPort(t *testing.T) {
	allurex.Run(t, cmUnitMeta("resolveCmHubUrl uses cmHubPort with trailing slash"), func(a *allurex.A) {
		a.Step("resolve", func() {
			cfg := config.FromMap(map[string]string{"cmHubPort": "4445"})
			require.Equal(t, "http://127.0.0.1:4445/", cfg.ResolveCmHubURL())
		})
	})
}

func TestConfigReader_ResolveCmHubURLRespectsCustomPort(t *testing.T) {
	allurex.Run(t, cmUnitMeta("resolveCmHubUrl respects custom cmHubPort"), func(a *allurex.A) {
		a.Step("resolve", func() {
			cfg := config.FromMap(map[string]string{"cmHubPort": "9444"})
			require.Equal(t, "http://127.0.0.1:9444/", cfg.ResolveCmHubURL())
		})
	})
}

func TestConfigReader_ResolveCmUiURLUsesCmUiPort(t *testing.T) {
	allurex.Run(t, cmUnitMeta("resolveCmUiUrl uses cmUiPort with trailing slash"), func(a *allurex.A) {
		a.Step("resolve", func() {
			cfg := config.FromMap(map[string]string{"cmUiPort": "8081"})
			require.Equal(t, "http://127.0.0.1:8081/", cfg.ResolveCmUiURL())
		})
	})
}

func TestConfigReader_ResolveCmUiURLRespectsCustomPort(t *testing.T) {
	allurex.Run(t, cmUnitMeta("resolveCmUiUrl respects custom cmUiPort"), func(a *allurex.A) {
		a.Step("resolve", func() {
			cfg := config.FromMap(map[string]string{"cmUiPort": "9080"})
			require.Equal(t, "http://127.0.0.1:9080/", cfg.ResolveCmUiURL())
		})
	})
}

func TestConfigReader_ResolveCmRemoteURLCombinesHubBaseAndWdHubPath(t *testing.T) {
	allurex.Run(t, cmUnitMeta("resolveCmRemoteUrl points WebDriver hub at CM-managed port"), func(a *allurex.A) {
		a.Step("resolve", func() {
			cfg := config.FromMap(map[string]string{"cmHubPort": "4445"})
			require.Equal(t, "http://127.0.0.1:4445/wd/hub", cfg.ResolveCmRemoteURL())
		})
	})
}

func cmUnitMeta(name string) allurex.Meta {
	return allurex.Meta{
		Name:      name,
		Package:   "config.ConfigReaderCmTest",
		Layer:     "unit",
		Component: "cm",
		Epic:      "cm",
		Suite:     "ConfigReader CM URLs",
		Tags:      []string{"unit", "cm"},
	}
}
